### DISCUSSION TITLE
Question trying to use JSON type with generic sdk / Apollo

### DISCUSSION THREAD

**Discussion by taschmidt** [dotansimha/graphql-code-generator#6146] [Q&A]
Hi there. Sorry if I'm being dumb here but I'm trying to accomplish something that seems like it would be fairly easy but I've been unsuccessful. I have to use the "generic sdk" plugin so I can control creating the Apollo client since we need custom authentication. My Requester is defined as follows:

```
const apolloRequester: Requester = async (doc, vars) => {
    const client = createApolloClient();
    const operationDefinition = doc.definitions.find(x => x.kind === 'OperationDefinition') as OperationDefinitionNode;
    switch (operationDefinition.operation) {
        case 'query': return (await client.query({ query: doc, variables: vars })).data;
        case 'mutation': return (await client.mutate({ mutation: doc, variables: vars })).data;
    }
};
```

I have a few fields that use AWSJSON (from AppSync) scalars which map to "any". The problem is these appear to go over the wire as simple strings. I would love to have a wrapper that automatically does a JSON parse/stringify but I just can't figure out where to plug this in.

UPDATE: I should also mention I'm attempting to do this from a backend Node component.

Any help would be greatly appreciated!!

Thanks,
tim

**Comment by dotansimha** [comment]
You don't have to use that setup for adding custom logic to Apollo, you can use Apollo-Link for that.

**Reply by taschmidt** [reply]
Can you elaborate on that? I'm not super familiar with Apollo either. I copied code from another dev which does, in fact, use Apollo link to handle the IAM auth. I'm assuming you're proposing that I would use the standard TS React Apollo plugin but how do I integrate the Link portion?

**Reply by taschmidt** [reply]
Ok, I got it working without the generic sdk but the react-apollo plugin seemed extraneous since I can't use hooks server side anyway. I ended up using the "typed-document-node" to generate my doc nodes so I can do the following:

```
createApolloClient().query({
    query: AssetDocument,
    variables: { guid: '60B148DC-5C66-441D-88FE-98A8E27B2888', assetID: 'zviENg-5prl14xH5DcK0p', revision: 5 }
}).then(console.dir);
```

This works but I'm still back to my original problem. JSON properties still come back as strings. So how can I automatically parse these on output and stringify on input?

**Reply by dotansimha** [reply]
> Ok, I got it working without the generic sdk but the react-apollo plugin seemed extraneous since I can't use hooks server side anyway. I ended up using the "typed-document-node" to generate my doc nodes so I can do the following:
> 
> ```
> createApolloClient().query({
>     query: AssetDocument,
>     variables: { guid: '60B148DC-5C66-441D-88FE-98A8E27B2888', assetID: 'zviENg-5prl14xH5DcK0p', revision: 5 }
> }).then(console.dir);
> ```
> 
> This works but I'm still back to my original problem. JSON properties still come back as strings. So how can I automatically parse these on output and stringify on input?

I'm really not sure, this sounds like a specific topic regarding NextJS setup, no? I'm not sure it's related to codegen.

**Reply by taschmidt** [reply]
I'm not using NextJS, did you mean this might be something to configure with the Apollo client? That's actually where I started looking but I really wasn't sure how exactly Apollo rehydrated returned objects. Thanks anyway!

**Reply by dotansimha** [reply]
Really not sure :X Sounds like something that worth checking with Apollo, since the code generated by codegen it just a wrapper on top of it, and not a runtime code.
